FamiliyDoctor/FamiliyDoctor/src/main/java/com/familydoctor/services/DoctorServiceImpl.java

```diff
@@ -0,0 +1,141 @@
+        this.patientRepository = patientRepository;
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public List<DoctorDTO> getAllDoctors() {
+        List<Doctor> doctors = doctorRepository.findAll();
+        return mapDoctorsToDTOs(doctors);
+    }
+
+    @Override
+    public DoctorDTO getDoctorById(Long id) {
+        Doctor doctor = doctorRepository.findById(id)
+                .orElseThrow(() -> new DoctorNotFoundException("Doctor not found: Doctor with id " + id, id));
+        return mapDoctorToDTO(doctor);
+    }
+
+    @Override
+    public DoctorDTO createDoctor(DoctorDTO doctorDTO) {
+        Doctor doctorToSave = mapDTOToDoctor(doctorDTO);
+        Doctor savedDoctor = doctorRepository.save(doctorToSave);
+        return mapDoctorToDTO(savedDoctor);
+    }
+
+    @Override
+    public DoctorDTO updateDoctor(Long id, DoctorDTO doctorDTO) {
+        Doctor existingDoctor = doctorRepository.findById(id)
+                .orElseThrow(() -> new DoctorNotFoundException("Doctor not found: Doctor with id " + id, id));
+        Doctor updatedDoctor = mapDTOToDoctor(doctorDTO);
+        updatedDoctor.setId(existingDoctor.getId());
+        Doctor savedDoctor = doctorRepository.save(updatedDoctor);
+        return mapDoctorToDTO(savedDoctor);
+    }
+
+    @Override
+    public void deleteDoctor(Long id) {
+        if (!doctorRepository.existsById(id)) {
+            throw new DoctorNotFoundException("Doctor not found: Doctor with id " + id, id);
+        }
+        doctorRepository.deleteById(id);
+    }
+    @Override
+    public PatientDTO addPatient(Long doctorId, PatientDTO patientDTO) {
+        Doctor doctor = doctorRepository.findById(doctorId)
+                .orElseThrow(() -> new DoctorNotFoundException("Doctor not found: Doctor with id " + doctorId, doctorId));
+
+        Patient patient = new Patient();
+        patient.setFirstName(patientDTO.getFirstName());
+        patient.setLastName(patientDTO.getLastName());
+        patient.setEmail(patientDTO.getEmail());
+        patient.setAddress(patientDTO.getAddress());
+
+        Random random = new Random();
+        long id = random.nextLong();
+        patient.setId(id);
+
+        doctor.getPatients().add(patient);
+        doctorRepository.save(doctor);
+        return mapPatientToDTO(patient);
+    }
+    @Override
+    public List<PatientDTO> getPatientsByAge(Long doctorId, int age) {
+        Doctor doctor = doctorRepository.findById(doctorId)
+                .orElseThrow(() -> new DoctorNotFoundException("Doctor not found: Doctor with id " + doctorId, doctorId));
+        List<Patient> patients = new ArrayList<>();
+        for (Patient patient : patients) {
+            int patientAge = calculateAge(patient.getDateOfBirth());
+            if (patientAge < 18) {
+                System.out.println("Patient is under age");
+            } else {
+                System.out.println("Patient is an adult");
```

Comment:
Please remove all System.out.println methods from the application. We use loggers if we want to print some information/errors in the console. Also, I don't see the point of this if-else structure. If we don't allow patients under 18 in our application, this should be a validation only when we are creating a patient, and after that, we will know for sure that all the patients are 18+. And we never do some logic to just print something in the console, but rather throw an exception maybe and we try to tell some relevant information to the API user.